docs/decisions/0030-branching-strategy.md

```diff
@@ -0,0 +1,125 @@
+
+<img src="./diagrams/git-flow.png" alt="GitFlow" width="700"/>
+
+Pros:
+- Clear separation between code under development and production-ready code.
+- Efficient release management.
+
+Cons:
+- More complex than GitHub Flow, which may be overwhelming for smaller teams or projects that do not require as much structure.
+- Less suited for projects that prioritize continuous deployment, as it emphasizes a more controlled release process.
+- Not ideal for projects with continuous deployment due to the overhead of managing multiple branches.
+- Spaghetti history in Git - [GitFlow considered harmful](https://www.endoflineblog.com/gitflow-considered-harmful)
+
+# SK branching strategies
+Today, the SK SDK is available in three languages: .NET, Java and Python. All of them coexist in the same Git repository, organized under corresponding folders. However, the branching strategies for those differ.
+
+For both .NET and Python versions, development takes place in short-lived topic branches that branch off the 'main' branch. These topic branches are merged back into the 'main' branch when features are considered production-ready through PR reviews, unit tests, and integration test runs. Releases are carried out directly from the 'main' branch. This approach aligns with the GitHub Flow branching strategy, with a minor deviation where releases are conducted weekly rather than being continuously deployed.
+
+The Java version of SK adheres to the Git-Flow strategy by being developed in a dedicated development branch. Topic branches are created from the development branch and merged back through pull requests after unit tests and integration test runs. Release branches are also created from the development branch and merged to both the development branch and the 'main' one when a release is considered production-ready. This strategy deviates slightly from vanilla Git-Flow in that release artifacts are generated from release branches rather than from the 'main' branch.
+
+## Decision Drivers  
+- The strategy should be easy to implement and maintain without requiring significant investments.
+- The strategy should allow for maintaining several releases in parallel if required.
+- Ideally, the strategy is intuitive and simple so that everyone familiar with Git can adopt and follow it.
+- Ideally, all SK languages are able to adopt and use the same branching strategy.
```

Comment:
Some additional drivers:

- Ability to continually deploy new release with minimal overhead
- Ability to release language versions independently and on different schedules
- Allow the .Net, Java and Python teams to be able to operate independantly
- Ability to patch a release (for all languages)
- Consolidation of PR's and Issues to simplify the triage and review process

Also is there a section which explains the current strategy (and associated issues) being used by the Java team?

Reply:
Thanks, added.

Java strategy is explained in the "SK branching strategies" section:
<img width="769" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/68852919/3bb22b64-7e32-47cb-a35d-350fe81aa891">
The problem with squash merges is mentioned in the "Branch per SK language" option.